Clearer error message when using the `ro-crate-metadata.json` rather than the crate directory as input

We are using the validator at the ELIXIR BioHackathon (hence the influx of new issues!).

When someone uses the validator for the first time, they will usually use `path/to/crate/ro-crate-metadata.json` as input rather than just `path/to/crate/`. This gives an error about an invalid path which isn't clear how to resolve:

```
ROCrateInvalidURIError: Invalid URI "PosixPath('tests/data/crates/valid/workflow-roc/ro-crate-metadata.json')"


[ERROR] 

[FAILED] Unexpected error: Invalid URI "PosixPath('tests/data/crates/valid/workflow-roc/ro-crate-metadata.json')" !!!
```

There should be a clearer error message (and documentation!) explaining that the user should pass the whole directory as input, not the `ro-crate-metadata.json`.

--- a/rocrate_validator/cli/commands/errors.py
+++ b/rocrate_validator/cli/commands/errors.py
@@ -18,7 +18,8 @@ import textwrap
 from rich.console import Console
 
 import rocrate_validator.log as logging
-from rocrate_validator.errors import InvalidProfilePath, ProfileNotFound, ProfilesDirectoryNotFound
+from rocrate_validator.errors import (InvalidProfilePath, ProfileNotFound,
+                                      ProfilesDirectoryNotFound)
 
 # Create a logger for this module
 logger = logging.getLogger(__name__)

--- a/rocrate_validator/cli/commands/validate.py
+++ b/rocrate_validator/cli/commands/validate.py
@@ -37,10 +37,12 @@ from rocrate_validator.cli.commands.errors import handle_error
 from rocrate_validator.cli.main import cli, click
 from rocrate_validator.cli.utils import Console, get_app_header_rule
 from rocrate_validator.colors import get_severity_color
+from rocrate_validator.errors import ROCrateInvalidURIError
 from rocrate_validator.events import Event, EventType, Subscriber
 from rocrate_validator.models import (LevelCollection, Profile, Severity,
                                       ValidationResult)
-from rocrate_validator.utils import URI, get_profiles_path
+from rocrate_validator.utils import (URI, get_profiles_path,
+                                     validate_rocrate_uri)
 
 # from rich.markdown import Markdown
 # from rich.table import Table
@@ -58,17 +60,11 @@ def validate_uri(ctx, param, value):
     """
     if value:
         try:
-            # parse the value to extract the scheme
-            uri = URI(value)
-            if not uri.is_remote_resource() and not uri.is_local_directory() and not uri.is_local_file():
-                raise click.BadParameter(f"Invalid RO-Crate URI \"{value}\": "
-                                         "it MUST be a local directory or a ZIP file (local or remote).", param=param)
-            if not uri.is_available():
-                raise click.BadParameter("RO-crate URI not available", param=param)
-        except ValueError as e:
-            logger.debug(e)
-            raise click.BadParameter("Invalid RO-crate path or URI", param=param)
-
+            validate_rocrate_uri(value)
+        except ROCrateInvalidURIError as e:
+            if logger.isEnabledFor(logging.DEBUG):
+                logger.exception(e)
+            raise click.BadParameter(e.message, param=param)
     return value
 
 

--- a/rocrate_validator/errors.py
+++ b/rocrate_validator/errors.py
@@ -243,9 +243,9 @@ class CheckValidationError(ValidationError):
 class ROCrateInvalidURIError(ROCValidatorError):
     """Raised when an invalid URI is provided."""
 
-    def __init__(self, uri: Optional[str] = None, message: Optional[str] = None):
+    def __init__(self, uri: str, message: Optional[str] = None):
         self._uri = uri
-        self._message = message
+        self._message = message or self.default_error_message(uri)
 
     @property
     def uri(self) -> Optional[str]:
@@ -258,13 +258,15 @@ class ROCrateInvalidURIError(ROCValidatorError):
         return self._message
 
     def __str__(self) -> str:
-        if self._message:
-            return f"Invalid URI \"{self._uri!r}\": {self._message!r}"
-        else:
-            return f"Invalid URI \"{self._uri!r}\""
+        return self._message
 
     def __repr__(self):
         return f"ROCrateInvalidURIError({self._uri!r})"
+
+    @classmethod
+    def default_error_message(cls, uri: str) -> str:
+        return f"Invalid RO-Crate URI \"{uri}\": "\
+            "it MUST be the local path to the root RO-Crate directory or a ZIP file (local or remote)."
 
 
 class ROCrateMetadataNotFoundError(ROCValidatorError):

--- a/rocrate_validator/errors.py
+++ b/rocrate_validator/errors.py
@@ -265,8 +265,8 @@ class ROCrateInvalidURIError(ROCValidatorError):
 
     @classmethod
     def default_error_message(cls, uri: str) -> str:
-        return f"Invalid RO-Crate URI \"{uri}\": "\
-            "it MUST be the local path to the root RO-Crate directory or a ZIP file (local or remote)."
+        return f"\"{uri}\" is not a valid RO-Crate URI. "\
+            "It MUST be either a local path to the RO-Crate root directory or a local/remote RO-Crate ZIP file."
 
 
 class ROCrateMetadataNotFoundError(ROCValidatorError):

--- a/rocrate_validator/rocrate.py
+++ b/rocrate_validator/rocrate.py
@@ -362,8 +362,8 @@ class ROCrate(ABC):
     @staticmethod
     def new_instance(uri: Union[str, Path, URI]) -> 'ROCrate':
         # check if the URI is valid
-        if not uri:
-            raise ValueError("Invalid URI")
+        validate_rocrate_uri(uri, silent=False)
+        # create a new instance based on the URI
         if not isinstance(uri, URI):
             uri = URI(uri)
         # check if the URI is a local directory
@@ -376,7 +376,7 @@ class ROCrate(ABC):
         if uri.is_remote_resource():
             return ROCrateRemoteZip(uri)
         # if the URI is not supported, raise an error
-        raise ROCrateInvalidURIError(uri=uri, message="Unsupported URI")
+        raise ROCrateInvalidURIError(uri=uri, message="Unsupported RO-Crate URI")
 
 
 class ROCrateLocalFolder(ROCrate):

--- a/rocrate_validator/rocrate.py
+++ b/rocrate_validator/rocrate.py
@@ -27,7 +27,7 @@ from rdflib import Graph
 
 from rocrate_validator import log as logging
 from rocrate_validator.errors import ROCrateInvalidURIError
-from rocrate_validator.utils import URI
+from rocrate_validator.utils import URI, validate_rocrate_uri
 
 # set up logging
 logger = logging.getLogger(__name__)

--- a/rocrate_validator/utils.py
+++ b/rocrate_validator/utils.py
@@ -480,7 +480,7 @@ def validate_rocrate_uri(uri: Union[str, URI], silent: bool = False) -> bool:
                 raise errors.ROCrateInvalidURIError(uri)
             # check if the resource is available
             if not uri.is_available():
-                raise errors.ROCrateInvalidURIError(uri, message=f"RO-crate URI \"{uri}\" not available")
+                raise errors.ROCrateInvalidURIError(uri, message=f"The RO-crate at the URI \"{uri}\" is not available")
             return True
         except ValueError as e:
             logger.error(e)

--- a/rocrate_validator/utils.py
+++ b/rocrate_validator/utils.py
@@ -458,6 +458,43 @@ class URI:
         return hash(self._uri)
 
 
+def validate_rocrate_uri(uri: Union[str, URI], silent: bool = False) -> bool:
+    """
+    Validate the RO-Crate URI
+
+    :param uri: The RO-Crate URI
+    :param silent: If True, do not raise an exception
+    :return: True if the URI is valid, False otherwise
+    """
+    try:
+        assert uri, "The RO-Crate URI is required"
+        assert isinstance(uri, (str, URI)), "The RO-Crate URI must be a string or URI object"
+        error_message = f"Invalid RO-Crate URI \"{uri}\": "\
+            "it MUST be the local path to the root RO-Crate directory or a ZIP file (local or remote)."
+        try:
+            # parse the value to extract the scheme
+            uri = URI(uri) if isinstance(uri, str) else uri
+            # check if the URI is a remote resource or local directory or local file
+            if not uri.is_remote_resource() and not uri.is_local_directory() and not uri.is_local_file():
+                raise errors.ROCrateInvalidURIError(error_message)
+            # check if the local file is a ZIP file
+            if uri.is_local_file() and uri.as_path().suffix != ".zip":
+                raise errors.ROCrateInvalidURIError(error_message)
+            # check if the resource is available
+            if not uri.is_available():
+                raise errors.ROCrateInvalidURIError(f"RO-crate URI \"{uri}\" not available")
+            return True
+        except ValueError as e:
+            logger.error(e)
+            if logger.isEnabledFor(logging.DEBUG):
+                logger.exception(e)
+            raise errors.ROCrateInvalidURIError(error_message)
+    except Exception as e:
+        if not silent:
+            raise e
+        return False
+
+
 class MapIndex:
 
     def __init__(self, name: str, unique: bool = False):

--- a/rocrate_validator/utils.py
+++ b/rocrate_validator/utils.py
@@ -375,7 +375,7 @@ class URI:
         except Exception as e:
             if logger.isEnabledFor(logging.DEBUG):
                 logger.debug(e)
-            raise ValueError("Invalid URI: %s", uri)
+            raise ValueError("Invalid URI: %s" % uri)
 
     @property
     def uri(self) -> str:

--- a/rocrate_validator/utils.py
+++ b/rocrate_validator/utils.py
@@ -469,26 +469,24 @@ def validate_rocrate_uri(uri: Union[str, URI], silent: bool = False) -> bool:
     try:
         assert uri, "The RO-Crate URI is required"
         assert isinstance(uri, (str, URI)), "The RO-Crate URI must be a string or URI object"
-        error_message = f"Invalid RO-Crate URI \"{uri}\": "\
-            "it MUST be the local path to the root RO-Crate directory or a ZIP file (local or remote)."
         try:
             # parse the value to extract the scheme
             uri = URI(uri) if isinstance(uri, str) else uri
             # check if the URI is a remote resource or local directory or local file
             if not uri.is_remote_resource() and not uri.is_local_directory() and not uri.is_local_file():
-                raise errors.ROCrateInvalidURIError(error_message)
+                raise errors.ROCrateInvalidURIError(uri)
             # check if the local file is a ZIP file
             if uri.is_local_file() and uri.as_path().suffix != ".zip":
-                raise errors.ROCrateInvalidURIError(error_message)
+                raise errors.ROCrateInvalidURIError(uri)
             # check if the resource is available
             if not uri.is_available():
-                raise errors.ROCrateInvalidURIError(f"RO-crate URI \"{uri}\" not available")
+                raise errors.ROCrateInvalidURIError(uri, message=f"RO-crate URI \"{uri}\" not available")
             return True
         except ValueError as e:
             logger.error(e)
             if logger.isEnabledFor(logging.DEBUG):
                 logger.exception(e)
-            raise errors.ROCrateInvalidURIError(error_message)
+            raise errors.ROCrateInvalidURIError(uri)
     except Exception as e:
         if not silent:
             raise e

--- a/tests/unit/test_uri.py
+++ b/tests/unit/test_uri.py
@@ -16,7 +16,9 @@ import unittest
 
 import pytest
 
-from rocrate_validator.utils import URI
+from rocrate_validator.errors import ROCrateInvalidURIError
+from rocrate_validator.utils import URI, validate_rocrate_uri
+from tests.ro_crates import ValidROC
 
 
 def test_valid_url():
@@ -80,6 +82,69 @@ def test_path_with_fragment():
 def test_path_without_fragment():
     uri = URI("/path/to/file.txt")
     assert uri.fragment is None
+
+
+def test_rocrate_uri_local_folder_valid():
+    uri = URI(ValidROC().workflow_roc)
+    assert validate_rocrate_uri(uri), f"The URI {uri} should be valid"
+
+
+def test_rocrate_uri_local_folder_invalid():
+    # Test with a non-existent folder
+    uri = URI("path/to/folder")
+    # Use silent mode to avoid printing the error message
+    assert not validate_rocrate_uri(uri, silent=True), f"The URI {uri} should be invalid"
+
+    # Use verbose mode to print the error message
+    with pytest.raises(ROCrateInvalidURIError) as excinfo:
+        validate_rocrate_uri(uri, silent=False)
+    assert str(
+        excinfo.value) == f"\"{uri}\" is not a valid RO-Crate URI. "\
+        "It MUST be either a local path to the RO-Crate root directory "\
+        "or a local/remote RO-Crate ZIP file."
+
+
+def test_rocrate_uri_local_zip_valid():
+    uri = URI(ValidROC().sort_and_change_archive)
+    assert validate_rocrate_uri(uri), f"The URI {uri} should be valid"
+
+
+def test_rocrate_uri_local_zip_invalid():
+    # Test with a non-existent zip file
+    uri = URI("path/to/zipfile.zip")
+    # Use silent mode to avoid printing the error message
+    assert not validate_rocrate_uri(uri, silent=True), f"The URI {uri} should be invalid"
+
+    # Use verbose mode to print the error message
+    with pytest.raises(ROCrateInvalidURIError) as excinfo:
+        validate_rocrate_uri(uri, silent=False)
+    assert str(
+        excinfo.value) == f"\"{uri}\" is not a valid RO-Crate URI. "\
+        "It MUST be either a local path to the RO-Crate root directory "\
+        "or a local/remote RO-Crate ZIP file."
+
+
+def test_rocrate_uri_remote_valid():
+    uri = URI(ValidROC().sort_and_change_remote)
+    assert validate_rocrate_uri(uri), f"The URI {uri} should be valid"
+
+
+def test_rocrate_uri_remote_invalid():
+
+    with pytest.raises(ValueError) as excinfo:
+        URI("httpx:///example.com")
+    assert str(excinfo.value) == "Invalid URI: httpx:///example.com"
+
+    # Test with an invalid remote URL
+    uri = URI("https:///example.com")
+    # Use silent mode to avoid printing the error message
+    assert not validate_rocrate_uri(uri, silent=True), f"The URI {uri} should be invalid"
+
+    # Use verbose mode to print the error message
+    with pytest.raises(ROCrateInvalidURIError) as excinfo:
+        validate_rocrate_uri(uri, silent=False)
+    assert str(
+        excinfo.value) == f"The RO-crate at the URI \"{uri}\" is not available"
 
 
 if __name__ == '__main__':

